Fix custom function forward AD internal assert

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #71569
* __->__ #71531
* #71530

Based on the comment above the original internal assert, this is the desired check.
1. Don't error, and automatically make jvp return a view for that tensor output (this is easier than I originally thought: https://github.com/pytorch/pytorch/pull/71531#discussion_r789211877)
2. Error (currently doing)

Differential Revision: [D33695399](https://our.internmc.facebook.com/intern/diff/D33695399)

soulitzer committed:
Fix custom function forward AD internal assert

[ghstack-poisoned]

albanD reviewed (commented):
change makes sense, we should have a test

soulitzer committed:
Update on "Fix custom function forward AD internal assert"


Based on the comment above the original internal assert, this is the desired check.

Doing no-op in forward will still fail because it gets turned into a view in post processing. We might want to fix this.
1. Detect this case and automatically make jvp return a view for that tensor output
2. Error

[ghstack-poisoned]

soulitzer commented on torch/csrc/autograd/custom_function.cpp:
```diff
@@ -260,6 +261,7 @@ optional_variable_list _process_backward_mode_ad(
       // Run in no_grad mode to mimic the behavior of the forward.
       {
         AutoGradMode grad_mode(false);
+        at::AutoFwGradMode fw_grad_mode(false);
```
It turns out that this particular output already has the forward grad we desired!

So... an alternate solution is actually to not have this guard, but if we detect `is_input && !mark_dirty`, then we simply continue instead of setting fwgrad.

albanD replied:
Yes that sounds like the right thing to do!

soulitzer committed:
Update on "Fix custom function forward AD internal assert"


Based on the comment above the original internal assert, this is the desired check.
1. Don't error, and automatically make jvp return a view for that tensor output (this is easier than I originally thought: https://github.com/pytorch/pytorch/pull/71531#discussion_r789211877)
2. Error (currently doing)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33695399).

soulitzer committed:
Update on "Fix custom function forward AD internal assert"


Based on the comment above the original internal assert, this is the desired check.
1. Don't error, and automatically make jvp return a view for that tensor output (this is easier than I originally thought: https://github.com/pytorch/pytorch/pull/71531#discussion_r789211877)
2. Error (currently doing)

Differential Revision: [D33695399](https://our.internmc.facebook.com/intern/diff/D33695399)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33695399).

albanD reviewed (approved):
Thanks!

soulitzer committed:
Update on "Fix custom function forward AD internal assert"


Based on the comment above the original internal assert, this is the desired check.
1. Don't error, and automatically make jvp return a view for that tensor output (this is easier than I originally thought: https://github.com/pytorch/pytorch/pull/71531#discussion_r789211877)
2. Error (currently doing)

Differential Revision: [D33695399](https://our.internmc.facebook.com/intern/diff/D33695399)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33695399).

soulitzer committed:
Update on "Fix custom function forward AD internal assert"


Based on the comment above the original internal assert, this is the desired check.
1. Don't error, and automatically make jvp return a view for that tensor output (this is easier than I originally thought: https://github.com/pytorch/pytorch/pull/71531#discussion_r789211877)
2. Error (currently doing)

Differential Revision: [D33695399](https://our.internmc.facebook.com/intern/diff/D33695399)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33695399).
